# Data Cleaning with Pandas

In this notebook we'll go through a few basic data cleaning steps that should be performed on all new datasets where necessary.

We'll go through the process with both the `orders` and `orderlines` datasets. You can then practice these skills by cleaning the `products` dataset yourself

In [1]:
import pandas as pd

In [2]:
pd.set_option('display.max_colwidth', None) # to increase the width of the columns

In [ ]:
# orders.csv
url = "https://drive.google.com/file/d/1Vu0q91qZw6lqhIqbjoXYvYAQTmVHh6uZ/view?usp=sharing"
path = "https://drive.google.com/uc?export=download&id="+url.split("/")[-2]
orders = pd.read_csv(path)

# orderlines.csv
url = "https://drive.google.com/file/d/1FYhN_2AzTBFuWcfHaRuKcuCE6CWXsWtG/view?usp=sharing"
path = "https://drive.google.com/uc?export=download&id="+url.split("/")[-2]
orderlines = pd.read_csv(path)

Before we begin, let's create a copy of the `orders` and `orderlines` DataFrames. This way we are sure any of our changes won't affect the original DataFrames.

In [ ]:
orders_df = orders.copy()

In [ ]:
orderlines_df = orderlines.copy()

One of the best ways to begin data cleaning is by exploring using `.info()`. This will tell us:
* The shape of the DataFrame
* The names of the columns
* If there are any missing values
* The datatypes of the columns

By exploring the missing values and correcting any incorrect datatypes, we often come across inconsistencies in our data.

Beyond this, we should also have a **check for any duplicate rows**.

Let's first deal with the duplicates, as it's nice and easy, then we'll explore what `.info()` has to tell us.

## 1.&nbsp; Duplicates
We can check for duplicates using the pandas [.duplicated()](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.duplicated.html) method.

We can then delete these rows, if we wish, using [.drop_duplicates()](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.drop_duplicates.html)

In [ ]:
# orders
orders_df.duplicated().sum()

0

In [ ]:
# orderlines
orderlines_df.duplicated().sum()

0

We have no duplicate rows in either DataFrame. Easy, there is no problem to solve. Normally though, if there were some duplicates, we'd drop the extra rows.

# 2.&nbsp; `.info()`

In [ ]:
orders_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 226909 entries, 0 to 226908
Data columns (total 4 columns):
 #   Column        Non-Null Count   Dtype  
---  ------        --------------   -----  
 0   order_id      226909 non-null  int64  
 1   created_date  226909 non-null  object 
 2   total_paid    226904 non-null  float64
 3   state         226909 non-null  object 
dtypes: float64(1), int64(1), object(2)
memory usage: 6.9+ MB


* `total_paid` has 5 missing values
* `created_date` should become datetime datatype

In [ ]:
orderlines_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 293983 entries, 0 to 293982
Data columns (total 7 columns):
 #   Column            Non-Null Count   Dtype 
---  ------            --------------   ----- 
 0   id                293983 non-null  int64 
 1   id_order          293983 non-null  int64 
 2   product_id        293983 non-null  int64 
 3   product_quantity  293983 non-null  int64 
 4   sku               293983 non-null  object
 5   unit_price        293983 non-null  object
 6   date              293983 non-null  object
dtypes: int64(4), object(3)
memory usage: 15.7+ MB


* `date` should be a datetime datatype
* `unit_price` should be a float datatype

## 3.&nbsp; Missing values

### 3.1.&nbsp; Orders
* `total_paid` has 5 missing values

In [ ]:
print(f"5 missing values represents {((orders_df.total_paid.isna().sum() / orders_df.shape[0])*100).round(5)}% of the rows in our DataFrame")

5 missing values represents 0.0022% of the rows in our DataFrame


A quick way to find out a percentage here, if you don't need to print out a sentence for yourself/students/colleagues is `.value_count(normalize=True)`

In [ ]:
orders_df.total_paid.isna().value_counts(normalize=True)

False    0.999978
True     0.000022
Name: total_paid, dtype: float64

As there is such a tiny amount of missing values, we will simply delete these rows, as we have enough data without them.

In [ ]:
orders_df = orders_df.loc[~orders.total_paid.isna(), :]

Should you have a significant number of missing values in the future, you have a choice:
+ you can impute the values
+ you can take the values from other DataFrames, if they are present there
+ you can delete the values
+ or any number of other creative solutions

Please, always consider how much time you have on your project, and what impact your method of choice will have on your final assesment.

### 3.2.&nbsp; Orderlines
There are no missing values in `orderlines`

## 4.&nbsp; Datatypes

### 4.1.&nbsp; Orders
* `created_date` should become datetime datatype

In [ ]:
orders_df["created_date"] = pd.to_datetime(orders_df["created_date"])

### 4.1.&nbsp; Orderlines
* `date` should be a datetime datatype
* `unit_price` should be a float datatype

#### 4.1.1.&nbsp; `date`

In [ ]:
orderlines_df["date"] = pd.to_datetime(orderlines_df["date"])

#### 4.1.2.&nbsp;`unit_price`

In [ ]:
orderlines_df["unit_price"] = pd.to_numeric(orderlines_df["unit_price"])

ValueError: ignored

As you can see when we try to convert `unit_price` to a numerical datatype, we receive a `ValueError` telling us that pandas doesn't understand the number `1.137.99`. This is probably because numbers cannot have 2 decimal points. Let's see if there are any other numbers like this.

In [ ]:
orderlines_df.unit_price.str.contains("\d+\.\d+\.\d+").value_counts()

False    257814
True      36169
Name: unit_price, dtype: int64

Looks like over 36000 rows in `orderlines` are affected by this problem. Let's work out how much that is as a percentage of our total data.

In [ ]:
two_dot_percentage = ((orderlines_df.unit_price.str.contains("\d+\.\d+\.\d+").value_counts()[1] / orderlines_df.shape[0])*100).round(2)
print(f"The 2 dot problem represents {two_dot_percentage}% of the rows in our DataFrame")

The 2 dot problem represents 12.3% of the rows in our DataFrame


This is a bit of a tricky decision as 12.3% is a significant amount of our data... and we might even end up losing a larger portion of our data than this too. For the moment we will delete the rows as we only have 2 weeks for this project and I'd like some quick, accurate results to show. If we have time at the end, we can come back and investigate this problem further, maybe there's a solution?

Each row of `orderlines` represents a product in an order. For example, if order number 175 contained 3 seperate products, then order 175 would have 3 rows in `orderlines`, one row for each of the products. If 2 of those products have 'normal' prices (14.99, 15.85) and 1 has a price with 2 decimal points (1.137.99), we need to remove the whole order and not just the affected row. If we only remove the row with 2 decimal places then any later analysis about products and prices could be misleading.

We therefore need to find the order numbers associated with the rows that have 2 decimal points, and then remove all the associated rows.

In [ ]:
two_dot_order_ids_list = orderlines_df.loc[orderlines_df.unit_price.str.contains("\d+\.\d+\.\d+"), "id_order"]

orderlines_df = orderlines_df.loc[~orderlines_df.id_order.isin(two_dot_order_ids_list)]

In [ ]:
orderlines_df.shape[0]

216250

We still have 216250 rows in orderlines to work with. This should be more than enough for our evaluation.

Now that all of the 2 decimal point prices have been removed, let's try again to convert the column `unit_price` to the correct datatype.

In [ ]:
orderlines_df["unit_price"] = pd.to_numeric(orderlines_df["unit_price"])

It worked perfectly

# Challenge: Clean the `products` DataFrame
Now it's your turn. Use the lessons you learnt above and clean the products DataFrame. You don't have to copy exactly what we did. Think about the consequences of your actions, sometimes it is ok to delete rows, other times you may wish to come up with more creative solutions.

In [45]:
# products.csv
url = "https://drive.google.com/file/d/1afxwDXfl-7cQ_qLwyDitfcCx3u7WMvkU/view?usp=sharing"
path = "https://drive.google.com/uc?export=download&id="+url.split("/")[-2]
products = pd.read_csv(path)

In [4]:
products_df = products.copy()

In [5]:
products_df.head()

sku                                           name  \
0  RAI0007              Silver Rain Design mStand Support   
1  APP0023              Apple Mac Keyboard Keypad Spanish   
2  APP0025               Mighty Mouse Apple Mouse for Mac   
3  APP0072  Apple Dock to USB Cable iPhone and iPod white   
4  KIN0007    Mac Memory Kingston 2GB 667MHz DDR2 SO-DIMM   

                                                            desc  price  \
0                   Aluminum support compatible with all MacBook  59.99   
1                      USB ultrathin keyboard Apple Mac Spanish.     59   
2                                         mouse Apple USB cable.     59   
3                          IPhone dock and USB Cable Apple iPod.     25   
4  2GB RAM Mac mini and iMac (2006/07) MacBook Pro (2006/07/08).  34.99   

  promo_price  in_stock      type  
0     499.899         1      8696  
1     589.996         0  13855401  
2     569.898         0      1387  
3     229.997         0      1230  
4       31.99         1      1364

In [6]:
products_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 19326 entries, 0 to 19325
Data columns (total 7 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   sku          19326 non-null  object
 1   name         19326 non-null  object
 2   desc         19319 non-null  object
 3   price        19280 non-null  object
 4   promo_price  19326 non-null  object
 5   in_stock     19326 non-null  int64 
 6   type         19276 non-null  object
dtypes: int64(1), object(6)
memory usage: 1.0+ MB


We'll go through the steps above in order
* Duplicates
* Missing values
* Datatypes

But I think we can all see straight away from `products.head()` above that some of the prices in `promo_price` look wrong. Let's make sure we deal with this later.

## Duplicates

In [7]:
products_df.duplicated().sum()

8746

Wow, that's a lot of duplicates. Let's get rid of them.

In [8]:
products_df = products_df.drop_duplicates().copy()

In [9]:
products_df

sku                                           name  \
0      RAI0007              Silver Rain Design mStand Support   
1      APP0023              Apple Mac Keyboard Keypad Spanish   
2      APP0025               Mighty Mouse Apple Mouse for Mac   
3      APP0072  Apple Dock to USB Cable iPhone and iPod white   
4      KIN0007    Mac Memory Kingston 2GB 667MHz DDR2 SO-DIMM   
...        ...                                            ...   
19321  BEL0376        Belkin Travel Support Apple Watch Black   
19322  THU0060   Enroute Thule 14L Backpack MacBook 13 "Black   
19323  THU0061    Enroute Thule 14L Backpack MacBook 13 "Blue   
19324  THU0062     Enroute Thule 14L Backpack MacBook 13 "Red   
19325  THU0063   Enroute Thule 14L Backpack MacBook 13 "Green   

                                                                                                desc  \
0                                                       Aluminum support compatible with all MacBook   
1                                                          USB ultrathin keyboard Apple Mac Spanish.   
2                                                                             mouse Apple USB cable.   
3                                                              IPhone dock and USB Cable Apple iPod.   
4                                      2GB RAM Mac mini and iMac (2006/07) MacBook Pro (2006/07/08).   
...                                                                                              ...   
19321                          compact and portable stand vertically or horizontally for Apple Watch   
19322  Backpack with capacity of 14 liter compartments MacBook up to 13 inches up to 10 inches Cases   
19323  Backpack with capacity of 14 liter compartments MacBook up to 13 inches up to 10 inches Cases   
19324  Backpack with capacity of 14 liter compartments MacBook up to 13 inches up to 10 inches Cases   
19325  Backpack with capacity of 14 liter compartments MacBook up to 13 inches up to 10 inches Cases   

       price promo_price  in_stock      type  
0      59.99     499.899         1      8696  
1         59     589.996         0  13855401  
2         59     569.898         0      1387  
3         25     229.997         0      1230  
4      34.99       31.99         1      1364  
...      ...         ...       ...       ...  
19321  29.99     269.903         1     12282  
19322  69.95     649.903         1      1392  
19323  69.95     649.903         1      1392  
19324  69.95     649.903         0      1392  
19325  69.95     649.903         1      1392  

[10580 rows x 7 columns]

## `.info()`

In [10]:
products_df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 10580 entries, 0 to 19325
Data columns (total 7 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   sku          10580 non-null  object
 1   name         10580 non-null  object
 2   desc         10573 non-null  object
 3   price        10534 non-null  object
 4   promo_price  10580 non-null  object
 5   in_stock     10580 non-null  int64 
 6   type         10530 non-null  object
dtypes: int64(1), object(6)
memory usage: 661.2+ KB


### Missing values
We can see from `.info()` above that we have missing values in `desc` and `price`

#### `desc`

In [11]:
products_df["desc"].isna().sum()

7

7 is a very small number to have missing, let's have a closer look

In [12]:
products_df.loc[products_df['desc'].isna(), :]

sku  \
16126  WDT0211-A   
16128  APP1622-A   
17843    PAC2334   
18152  KAN0034-A   
18490    HTE0025   
18612    OTT0200   
18690  HOW0001-A   

                                                                                name  \
16126                    Open - Purple 2TB WD 35 "PC Security Mac hard drive and NAS   
16128                      Open - Apple Smart Keyboard Pro Keyboard Folio iPad 9.7 "   
17843                                         Synology DS718 + NAS Server | 10GB RAM   
18152                       Open - Kanex USB-C Gigabit Ethernet Adapter MacBook 12 "   
18490                    Hyper Pearl 1600mAh battery Mini USB Mirror and Comic Blond   
18612                                OtterBox External Battery Power Pack 20000 mAHr   
18690  Open - Honeywell thermostat Lyric zonificador T6 Intelligent Wireless (cable)   

      desc      price promo_price  in_stock      type  
16126  NaN        107     814.659         0      1298  
16128  NaN  1.568.206   1.568.206         0      1298  
17843  NaN     566.35   5.659.896         0  12175397  
18152  NaN      29.99     237.925         0      1298  
18490  NaN      24.99       22.99         1      1515  
18612  NaN      79.99       56.99         1      1515  
18690  NaN     199.99   1.441.174         0  11905404

We have 2 choices here:
* We can quickly and easily remove these rows.
* Or, alternatively, the products names here are quite descriptive, so I'm tempted to just copy them to the description column, so that there is a description if we later want utilise this column. I wouldn't recommend this if this DataFrame was the source of truth for our website. But this is not the case here, and we're not faking any information (guessing a price or so), so I'm happy with this option

In [15]:
products_df.loc[products_df['desc'].isna(), 'desc'] = products_df.loc[products_df['desc'].isna(), 'name']

In [16]:
products_df.loc[products_df['desc'].isna(), :]

Empty DataFrame
Columns: [sku, name, desc, price, promo_price, in_stock, type]
Index: []

Did you also notice above that we have the dreaded two decimal point problem in both the `price` and `promo_price` columns? We can also see prices with 3 decimal places, prices should have 2 decimal places: this gives us more cause for concern

#### `price`

In [17]:
products_df.price.isna().sum()

46

In [18]:
print(f"The missing values in price are {(products_df.price.isna().value_counts(normalize=True)[1] * 100).round(2)}% of all rows in the DataFrame")

The missing values in price are 0.43% of all rows in the DataFrame


/var/folders/ts/3b69bt6s08sdw_hm__snx_n40000gn/T/ipykernel_36046/3940564348.py:1: FutureWarning: Series.__getitem__ treating keys as positions is deprecated. In a future version, integer keys will always be treated as labels (consistent with DataFrame behavior). To access a value by position, use `ser.iloc[pos]`
  print(f"The missing values in price are {(products_df.price.isna().value_counts(normalize=True)[1] * 100).round(2)}% of all rows in the DataFrame")


Let's simply delete these rows to ensure that we can trust the numbers in our final DataFrame. Afterall, the price is very important when investigating discounts.

Option 1: `.loc`

In [ ]:
products_df = products_df.loc[~products['price'].isna()]

Option 2: `.dropna()`

In [19]:
products_df = products_df.dropna(subset='price').copy()

### Data types

We saw from looking at the output of `.info()` that both `price` and `promo_price` have been stored as objects and not as a numerical datatypes. We also saw while solving other problems that both columns have some prices with 3 decimal places and others with 2 decimal points - the latter will prevent us from converting the datatype to numerical, so first we must investigate and solve these problems.

#### `price`

First, let's see how many values are affected by the 2-decimal-dot problems or 3 decimal places.

In [20]:
products_df.sample(50)

sku  \
14446  OTR0025-A   
1897   APP0017-A   
17130    IFX0170   
2006     PAC0713   
12698    OTT0143   
1152     OWC0119   
12677    MOP0082   
1087     SAM0067   
10975    GRT0391   
15856    BEZ0220   
12535    PUR0140   
3001     SAN0098   
15624    APP2011   
15187    MUJ0031   
16889    PAC1942   
12128    UAG0037   
15174    TWS0108   
2318     MOX0016   
536      SEV0028   
11902  LOG0142-A   
1189     DLK0083   
2569     JBL0097   
13678  TUC0190-A   
957      LEP0008   
19221    WAC0250   
1259     TUC0177   
18846    AP20067   
17113    SAN0182   
2450     LIF0063   
12968    OSM0007   
12701  MDT0007-A   
11351    SAM0093   
13020    CRU0053   
15846  WDT0246-A   
12692    OTT0137   
15277  PHI0048-A   
497      DRO0016   
17012    PAC1693   
15014    PAC1913   
527      MOP0027   
11593    LGE0030   
15827  NTE0109-A   
17567    AP20299   
13504    WAC0205   
13035    PAC1500   
16177    PAC2084   
14601    SAT0037   
6183     PAC1065   
2249     APP0976   
2938     MUV0167   

                                                                                                          name  \
14446                                                   Open - Startech Cable DVI-I Male to DVI-I (29-pin) 18m   
1897                                                              Open - Apple Keyboard Keypad for Mac Spanish   
17130                                           Battery iFixit MacBook Pro 13 "Retina (Late 2012 - Early 2013)   
2006   Samsung SSD 850 expansion kit PRO 512GB + 8GB RAM 1066Mhz + Datadoubler for iMac and Mac mini 2009 2009   
12698                                            Otterbox Clearly Protected gel Case iPhone 6 / 6S Transparent   
1152                                                             OWC ThunderBay 4 Thunderbolt 2 RAID 5 Edition   
12677                                                            Mophie Power Reserve 2600 mAh battery 1X Blue   
1087                                                                    External Recorder Samsung Bluray White   
10975                                          Griffin Survivor Case Tough Journey iPhone 6 / 6S Black / White   
15856                                     Be.ez LA robe Black Addicted MacBook Pro 13 "(Late 2016) Black / Red   
12535                                           Pure UltraSlim Case 03 + Protector iPhone 6 / 6S Plus Amarillo   
3001                                           SanDisk Connect Wireless Flash Drive 32GB Wireless Black / Gold   
15624                                  Apple Watch Series 1 42mm Aluminum Gold Case and Strap Sport Blue Night   
15187                                              Mujjo Single-Layered gloves color touch screen Black size M   
16889                                                  Apple Mac Pro 35GHz 6 cores | 64GB RAM | 256GB PCIe SSD   
12128                                                           Urban Armor Gear Case iPad Pro 97 "Red / Black   
15174                                                  Twelve South ActionSleeve Apple Watch Bracelet 38mm Red   
2318                                                         Moxie 2.5D Tempered glass iPhone 6s Plus / 6 Plus   
536                                                                    Budget Request repair Apple MacBook Pro   
11902                                                   (Open) Logitech Type + Keyboard Folio iPad Air 2 Black   
1189                                                                 D-Link Adapter DWA-125 Wireless USB 150 N   
2569                                                                             JBL GO Bluetooth Speaker Blue   
13678                                                           (Open) Tucano Folio Case Filo Black iPad Air 2   
957                                                                         Lepow Modre Wireless Speaker Green   
19221                                                         Education - Wacom Intuos Graphics Tablet S Black   
1259                        

In [22]:
price_problems_number = products_df.loc[(products_df.price.astype(str).str.contains(r"\d+\.\d+\.\d+"))|(products_df.price.astype(str).str.contains(r"\d+\.\d{3,}")), :].shape[0]
price_problems_number

542

In [23]:
print(f"The column price has in total {price_problems_number} wrong values. This is {round(((price_problems_number / products_df.shape[0]) * 100), 2)}% of the rows of the DataFrame")

The column price has in total 542 wrong values. This is 5.15% of the rows of the DataFrame


5.15% is a reasonable amount of our data. However, the price column will be important to understanding discounts, so I'd like it to be very trustworthy as we are basing business decisions on it. Therefore, we'll delete these rows

In [25]:
products_df = products_df.loc[(~products_df.price.astype(str).str.contains(r"\d+\.\d+\.\d+"))&(~products_df.price.astype(str).str.contains(r"\d+\.\d{3,}")), :]

In [26]:
products_df

sku                                           name  \
0      RAI0007              Silver Rain Design mStand Support   
1      APP0023              Apple Mac Keyboard Keypad Spanish   
2      APP0025               Mighty Mouse Apple Mouse for Mac   
3      APP0072  Apple Dock to USB Cable iPhone and iPod white   
4      KIN0007    Mac Memory Kingston 2GB 667MHz DDR2 SO-DIMM   
...        ...                                            ...   
19321  BEL0376        Belkin Travel Support Apple Watch Black   
19322  THU0060   Enroute Thule 14L Backpack MacBook 13 "Black   
19323  THU0061    Enroute Thule 14L Backpack MacBook 13 "Blue   
19324  THU0062     Enroute Thule 14L Backpack MacBook 13 "Red   
19325  THU0063   Enroute Thule 14L Backpack MacBook 13 "Green   

                                                                                                desc  \
0                                                       Aluminum support compatible with all MacBook   
1                                                          USB ultrathin keyboard Apple Mac Spanish.   
2                                                                             mouse Apple USB cable.   
3                                                              IPhone dock and USB Cable Apple iPod.   
4                                      2GB RAM Mac mini and iMac (2006/07) MacBook Pro (2006/07/08).   
...                                                                                              ...   
19321                          compact and portable stand vertically or horizontally for Apple Watch   
19322  Backpack with capacity of 14 liter compartments MacBook up to 13 inches up to 10 inches Cases   
19323  Backpack with capacity of 14 liter compartments MacBook up to 13 inches up to 10 inches Cases   
19324  Backpack with capacity of 14 liter compartments MacBook up to 13 inches up to 10 inches Cases   
19325  Backpack with capacity of 14 liter compartments MacBook up to 13 inches up to 10 inches Cases   

       price promo_price  in_stock      type  
0      59.99     499.899         1      8696  
1         59     589.996         0  13855401  
2         59     569.898         0      1387  
3         25     229.997         0      1230  
4      34.99       31.99         1      1364  
...      ...         ...       ...       ...  
19321  29.99     269.903         1     12282  
19322  69.95     649.903         1      1392  
19323  69.95     649.903         1      1392  
19324  69.95     649.903         0      1392  
19325  69.95     649.903         1      1392  

[9992 rows x 7 columns]

To complete our task, let's convert the column to a numeric datatype

In [27]:
products_df["price"] = pd.to_numeric(products_df["price"])

In [28]:
products_df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 9992 entries, 0 to 19325
Data columns (total 7 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   sku          9992 non-null   object 
 1   name         9992 non-null   object 
 2   desc         9992 non-null   object 
 3   price        9992 non-null   float64
 4   promo_price  9992 non-null   object 
 5   in_stock     9992 non-null   int64  
 6   type         9946 non-null   object 
dtypes: float64(1), int64(1), object(5)
memory usage: 624.5+ KB


#### `promo_price`

Again, let's begin by seeing how many values are affected by the 2-decimal-dots problem, or the 3 decimal-places problem

In [30]:
promo_problems_number = products_df.loc[(products_df.promo_price.astype(str).str.contains(r"\d+\.\d+\.\d+"))|(products_df.promo_price.astype(str).str.contains(r"\d+\.\d{3,}")), :].shape[0]
promo_problems_number

9232

In [31]:
print(f"The column promo_price has in total {promo_problems_number} wrong values. This is {round(((promo_problems_number / products_df.shape[0]) * 100), 2)}% of the rows of the DataFrame")

The column promo_price has in total 9232 wrong values. This is 92.39% of the rows of the DataFrame


WOW!!! That's a lot of wrong data. Let's have a quick investigate to check that's correct. We'll make a DataFrame by copy-pasting the code we used above and then look at a large sample to check that all the numbers in the `promo_price` column really have either 2 decimal points or 3 decimal places.

In [ ]:
promo_price_df = products_df.loc[(products_df.promo_price.astype(str).str.contains("\d+\.\d+\.\d+"))|(products_df.promo_price.astype(str).str.contains("\d+\.\d{3,}")), :]
promo_price_df.sample(50)

sku                                               name  \
14703    PAC1764                                   QNAP TS-431P NAS   
2696     WDT0288                  8GB SSHD 1TB WD Blue SSD 25 "SATA   
17484    APP2489                     Apple iPhone 64GB Space Gray X   
13648    GRT0447  Griffin Survivor iPhone Case Journey 8 Plus / ...   
11288    OWC0132            OWC Aura 2TB SSD for Apple Mac Pro 2013   
12389    PAC1424  Synology DS916 + Pack | 2GB RAM | WD 16TB Network   
19060    TRA0051              Transcend JetDrive PCIe SSD 825 240GB   
17267    AP20278  Like new - Apple Watch 42mm steel case and str...   
11559    PAC1270        Pack QNAP TS-251 | 4GB RAM | WD 4TB Network   
2291     MYK0013  MyKronoz ZeBracelet 2 receiving calls Bracelet...   
2309     LOG0158  Logitech AnyAngle iPad Mini Case with Stand Black   
14564    SAT0021             Support Satechi iPhone Aluminum Silver   
11723    OWC0182             OWC Mercury Electra 6G SSD Max 2TB 7mm   
1944     PAC0721  expansion kit 1TB SSD Crucial MX200 Mac mini a...   
18318    EVU0013  Evutec Aergo Ballistic Nylon Case + Support Ha...   
15748    APP2044  Open - Apple iPhone 7 Plus 32GB Silver - like new   
15742    APP2027   Open - Pink Apple iPhone 6s 16GB Gold - Like New   
931    CRU0016-2  Crucial memory Mac 8GB (2x4GB) SO-DIMM DDR3 10...   
14600    SAT0042                      Satechi Aluminum HDD Dock Mac   
15067  BNQ0037-A        Open - LED Monitor Benq GW2270 215 "DVI VGA   
18460    APP2691  Apple iMac Pro 27 "14-core Intel Xeon W 25GHz ...   
2288     TRA0033  Transcend StoreJet 100 2TB External Hard Drive...   
15844  OWC0124-A  Open - OWC HDD Installation Kit for iMac 27 "2...   
17879    PAC2371  Synology DS218 + NAS Server | 6GB RAM | 8TB (2...   
271      PAC0184  Apple MacBook Pro 133 "25GHz | RAM 16GB | 500G...   
18661    LAN0015                 LandingZone DOCK PRO 13 "Touch bar   
13601  BEL0240-A  (Open) Belkin InvisiGlass TrueClear invisible ...   
2317     MOX0015            Moxie 2.5D Tempered glass iPhone 6 / 6S   
11860    WDT0315              Blue WD 2TB Hard Drive 35 "Mac and PC   
2414     BEZ0181  Be.ez LA robe MacBook Air / Pro 13 Black / Orange   
1410     GOP0077                            GoPro Fetch Dog Harness   
1614     OLL0017  Olloclip 4-in-1 lens iPhone 6 / 6S and 6 / 6S ...   
11052    PAC1226  Pack QNAP TS-251 + | 2GB RAM | Seagate 4TB Iro...   
3153     XDO0021      X-Doria Lux Apple Watch Strap 38mm White Croc   
16699    APP2433  Apple Macbook 12 "Core i5 1.3GHz | 16GB | 512G...   
11331    UAG0027   Urban Armor Gear Pro 129 iPad Case "Blue / Black   
473      NTE0019  NewerTech NuPower battery MacBook Pro 15 "unib...   
15762    ELG0049         Dock Elgato Thunderbolt Hub USB-C 3 Silver   
18075    ADN0055                  Jot Adonith Dash 3 Pointer Bronze   
10692    PAC1778            Synology Pack DS216SE | Seagate 8TB NAS   
1442     BEA0025             Beats by Dr. Dre speaker Red Pill Case   
1681     PAC1796  Synology DS1815 Pack + | 8GB RAM | Seagate 48T...   
13422    APP1776  Smart Battery Apple Battery Case iPhone 6s Cha...   
2326     SPE0156  Speck CandyShell Grip Case for iPhone 6 Plus 5...   
2654     KAN0041          USB-C Kanex HDMI Adapter MacBook 12 "- 4K   
816      WDT0196    Western Digital My Book Thunderbolt Duo 6TB Mac   
1047     BNQ0019                BL2410PT LED Monitor Benq 24 "Black   
19216    WAC0257  Wacom Intuos Graphics Tablet M Bluetooth Pista...   
13148    SYN0158             Synology MailPlus License for 20 users   
16077    AP20145          Like new - Apple AirPort Time Capsule 3TB   

                                                    desc     price  \
14703  NAS with 16TB capacity Seagate Hard Drives for...   1021.95   
2696   Western Digital hard drive hybrid Mac and PC 1...     94.90   
17484                             New Free iPhone 64GB X   1159.00   
13648  rugged case with shock dispersion system for i...     29.99   
11288               2TB SSD hard drive fo

So we were correct, over 90% of the data in this column is corrupt. There's no point deleting all of these rows, then we would barely have a products table. Instead, as it's only this column that appears to be very untrustworthy, we will delete the column.

In [32]:
products_cl = products_df.drop(columns=["promo_price"])

In [33]:
products_cl.info()

<class 'pandas.core.frame.DataFrame'>
Index: 9992 entries, 0 to 19325
Data columns (total 6 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   sku       9992 non-null   object 
 1   name      9992 non-null   object 
 2   desc      9992 non-null   object 
 3   price     9992 non-null   float64
 4   in_stock  9992 non-null   int64  
 5   type      9946 non-null   object 
dtypes: float64(1), int64(1), object(4)
memory usage: 546.4+ KB


Obviously, there's now no need to convert `promo_price` to a numerical datatype

Don't forget to download/save your new DataFrames. Also, give them an obvious name, so that you know they are the cleaned version and not the original DataFrame.

In [34]:
# from google.colab import files

# orders_df.to_csv("orders_cl.csv", index=False)
# files.download("orders_cl.csv")

# orderlines_df.to_csv("orderlines_cl.csv", index=False)
# files.download("orderlines_cl.csv")

products_cl.to_csv("../data/products_cl.csv", index=False)
# files.download("products_cl.csv")

In [43]:
for p in products["price"].astype(str):
    if not isinstance(p, str):
        print(p)